File: 2023/day04/zig/solution.zig
const std = @import("std");

const Card = struct {
    winning: []u32,
    have: []u32,
};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // Read input file
    const content = try std.fs.cwd().readFileAlloc(allocator, "../input.txt", 1024 * 1024);
    defer allocator.free(content);

    // Trim trailing whitespace
    const trimmed = std.mem.trimRight(u8, content, " \t\n\r");

    // Collect lines using ArrayList (single pass)
    var lines: std.ArrayListUnmanaged([]const u8) = .empty;
    defer lines.deinit(allocator);

    var line_iter = std.mem.splitScalar(u8, trimmed, '\n');
    while (line_iter.next()) |line| {
        try lines.append(allocator, line);
    }

    // Part 1
    const part1_result = try part1(allocator, lines.items);
    std.debug.print("Part 1: {d}\n", .{part1_result});

    // Part 2
    const part2_result = try part2(allocator, lines.items);
    std.debug.print("Part 2: {d}\n", .{part2_result});
}

fn parseNumbers(allocator: std.mem.Allocator, text: []const u8) ![]u32 {
    var numbers: std.ArrayListUnmanaged(u32) = .empty;
    errdefer numbers.deinit(allocator);

    var iter = std.mem.tokenizeAny(u8, text, " ");
    while (iter.next()) |num_str| {
        try numbers.append(allocator, try std.fmt.parseInt(u32, num_str, 10));
    }

    return numbers.toOwnedSlice(allocator);
}

fn countMatches(winning: []const u32, have: []const u32) usize {
    var matches: usize = 0;
    for (have) |h| {
        for (winning) |w| {
            if (h == w) {
                matches += 1;
                break;
            }
        }
    }
    return matches;
}

fn parseCard(allocator: std.mem.Allocator, line: []const u8) !Card {
    // Split on ':'
    var colon_iter = std.mem.splitScalar(u8, line, ':');
    _ = colon_iter.next(); // Skip "Card X"
    const numbers_part = colon_iter.next() orelse return error.InvalidInput;

    // Split on '|'
    var pipe_iter = std.mem.splitScalar(u8, numbers_part, '|');
    const winning_part = pipe_iter.next() orelse return error.InvalidInput;
    const have_part = pipe_iter.next() orelse return error.InvalidInput;

    const winning = try parseNumbers(allocator, winning_part);
    const have = try parseNumbers(allocator, have_part);

    return Card{ .winning = winning, .have = have };
}

fn part1(allocator: std.mem.Allocator, lines: []const []const u8) !u64 {
    var total: u64 = 0;

    for (lines) |line| {
        const card = try parseCard(allocator, line);
        defer allocator.free(card.winning);
        defer allocator.free(card.have);

        const matches = countMatches(card.winning, card.have);
        if (matches > 0) {
            // Score is 2^(matches-1)
            total += @as(u64, 1) << @intCast(matches - 1);
        }
    }

    return total;
}

fn part2(allocator: std.mem.Allocator, lines: []const []const u8) !u64 {
    const card_count = lines.len;

    // Allocate array to track match counts for each card
    const matches = try allocator.alloc(usize, card_count);
    defer allocator.free(matches);

    // Calculate matches for each card
    for (lines, 0..) |line, i| {
        const card = try parseCard(allocator, line);
        defer allocator.free(card.winning);
        defer allocator.free(card.have);

        matches[i] = countMatches(card.winning, card.have);
    }

    // Track copies of each card (start with 1 of each)
    const copies = try allocator.alloc(u64, card_count);
    defer allocator.free(copies);
    @memset(copies, 1);

    // Process cards and cascade copies
    for (0..card_count) |i| {
        const match_count = matches[i];
        const current_copies = copies[i];

        // Win copies of the next match_count cards
        const end = @min(i + 1 + match_count, card_count);
        for (copies[i + 1 .. end]) |*c| {
            c.* += current_copies;
        }
    }

    // Sum all copies
    var total: u64 = 0;
    for (copies) |c| {
        total += c;
    }

    return total;
}

File: 2023/day04/rust/solution.rs
use std::collections::HashSet;
use std::fs;

struct Card {
    winning: HashSet<u32>,
    have: Vec<u32>,
}

fn parse_card(line: &str) -> Option<Card> {
    let (_, numbers) = line.split_once(':')?;
    let (winning_str, have_str) = numbers.split_once('|')?;

    let winning: HashSet<u32> = winning_str
        .split_whitespace()
        .filter_map(|s| s.parse().ok())
        .collect();

    let have: Vec<u32> = have_str
        .split_whitespace()
        .filter_map(|s| s.parse().ok())
        .collect();

    Some(Card { winning, have })
}

fn count_matches(card: &Card) -> usize {
    card.have.iter().filter(|n| card.winning.contains(n)).count()
}

fn part1(cards: &[Card]) -> u32 {
    cards.iter()
        .map(|card| {
            let matches = count_matches(card);
            if matches > 0 {
                1 << (matches - 1)  // 2^(matches-1)
            } else {
                0
            }
        })
        .sum()
}

fn part2(cards: &[Card]) -> u32 {
    let matches: Vec<usize> = cards.iter().map(count_matches).collect();
    let mut copies: Vec<u32> = vec![1; cards.len()];

    for (i, &m) in matches.iter().enumerate() {
        let current_copies = copies[i];
        for j in (i + 1)..(i + 1 + m).min(cards.len()) {
            copies[j] += current_copies;
        }
    }

    copies.iter().sum()
}

fn main() {
    // Read input file
    let input = fs::read_to_string("../input.txt")
        .expect("Failed to read input file");

    // Parse all cards
    let cards: Vec<Card> = input.lines()
        .filter_map(parse_card)
        .collect();

    // Solve both parts
    let part1_result = part1(&cards);
    let part2_result = part2(&cards);

    println!("Part 1: {}", part1_result);
    println!("Part 2: {}", part2_result);
}
